### DISCUSSION TITLE
Undefined class constant 'EM_AES_256' when installing laravel-backup

### DISCUSSION THREAD

**Discussion by eduardoarandah** [spatie/laravel-backup#1247] [Q&A]
# What happened

when I do `composer require spatie/laravel-backup` I get `Undefined class constant 'EM_AES_256'`

```
Generating optimized autoload files
ocramius/package-versions: Generating version class...
ocramius/package-versions: ...done generating version class
> Illuminate\Foundation\ComposerScripts::postAutoloadDump
> @php artisan package:discover --ansi

   Symfony\Component\Debug\Exception\FatalThrowableError  : Undefined class constant 'EM_AES_256'

  at /home/donors/tmp/donors/vendor/spatie/laravel-backup/config/backup.php:138
    134|         /*
    135|          * The encryption algorithm to be used for archive encryption.
    136|          * You can set it to `null` or `false` to disable encryption.
    137|          */
  > 138|         'encryption' => \ZipArchive::EM_AES_256,
    139|     ],
    140|
    141|     /*
    142|      * You can get notified when specific events occur. Out of the box you can use 'mail' and 'slack'.

  Exception trace:

  1   require()
      /home/donors/tmp/donors/vendor/laravel/framework/src/Illuminate/Support/ServiceProvider.php:64

  2   Illuminate\Support\ServiceProvider::mergeConfigFrom()
      /home/donors/tmp/donors/vendor/spatie/laravel-backup/src/BackupServiceProvider.php:38

  Please use the argument -v to see more details.
Script @php artisan package:discover --ansi handling the post-autoload-dump event returned with error code 1
```

# The environment

Ubuntu Server 
PHP 7.4.15
Composer version 2.0.9

# My guess

I guess it's something wrong about zip extension.

However it seems to be present:

```

php -i | grep zip
/etc/php74rc/conf.d/zip.ini
Registered PHP Streams => https, ftps, compress.zlib, php, file, glob, data, http, ftp, compress.bzip2, phar, zip
Registered Stream Filters => zlib.*, string.rot13, string.toupper, string.tolower, string.strip_tags, convert.*, consumed, dechunk, bzip2.*, convert.iconv.*
Stream Wrapper support => compress.bzip2://
Stream Filter support => bzip2.decompress, bzip2.compress
gzip compression => enabled
bzip2 compression => enabled
zip
Libzip version => 1.1.2

cat /etc/php74rc/conf.d/zip.ini
extension=zip.so

php -m | grep zip
zip

```

Now, check this out:

class exists:

```
php -r "echo \ZipArchive::class;"
ZipArchive
```

but constant doesn't:

```
php -r "echo \ZipArchive::EM_AES_256;"
PHP Fatal error:  Uncaught Error: Undefined class constant 'EM_AES_256' in Command line code:1
Stack trace:
#0 {main}
  thrown in Command line code on line 1
```

more info:

```
php -r "echo \ZipArchive::LIBZIP_VERSION;"
1.1.2
```

any ideas?

**Comment by eduardoarandah** [comment]
More info:

Here 

https://www.php.net/manual/en/zip.constants.php

Says 

ZipArchive::EM_AES_256 (int)
AES 256 encryption. Available as of PHP 7.2.0 and PECL zip 1.14.0, respectively, if built against libzip ≥ 1.2.0.

So this condition depends on how the extension was compiled. 

It's problematic because running composer dump-autoload in production breaks the system even if the feature is not used. 

I wonder if an isset(constant) in the package could work 🤔

**Comment by eduardoarandah** [comment]
Constant introduced in:
https://github.com/spatie/laravel-backup/blob/6.15.0/config/backup.php#L138

Not present in:
https://github.com/spatie/laravel-backup/blob/6.14.4/config/backup.php#L126

I made a if in a fork of 6.15.0 with the fix
https://github.com/spatie/laravel-backup/compare/6.15.0...eduardoarandah:fix-undefined-constant?expand=1

(but I don't know how to PR into 6.15.1 or something like that)

**Comment by eduardoarandah** [comment]
This was problematic because locally I made composer update and it worked. 

Tests passed. 

Then CI/CD worked 

But production broke the entire app because autoload couldn’t be generated. 
And error wasn’t clear at all because composer didn’t point the exact point it happened.

**Comment by nabeelio** [comment]
Running into the same issue with customers on a production instance in cPanel. It just needs to be rebuilt? I'll try that out and report back.

**Reply by nabeelio** [reply]
Ok, this is broken on cPanel v94.0.2 (the latest) under CentOS 7.9, which is a common configuration. I have the latest kernel and OS packages, recompiled PHP 7.3+, and the zip extension. 

@freekmurze I think the issue #1244 needs to be reopened and an isset or something used in the default since that constant won't be present everywhere.

**Reply by nabeelio** [reply]
Actually yes there needs to be a fix in the base library - it can't be overridden it seems because Laravel is first calling mergeFrom() with the base config file before it looks to be loading it from the `/config` directory. So it bombs out before you can even set that value to false

**Reply by nabeelio** [reply]
Same error in the `php:7.4-fpm-alpine` Docker container

**Comment by eduardoarandah** [comment]
If you got here with same problem, **temporary** solution:

in your composer.json use `"spatie/laravel-backup": "6.14.4",`

**Comment by wilsenhc** [comment]
Hello, @freekmurze can the fix for #1254 be backported to v6 with support for PHP 7.3?

**Reply by freekmurze** [reply]
Feel free to send a PR for that
